GH-38998: [Java] Build memory-core and memory-unsafe as JPMS modules

jduo opened the pull request:
Depends on #39134 

### Rationale for this change
Part of modularizing Arrow. Allows these components to be used by tools that require JPMS modules such as jlink.

### What changes are included in this PR?
- Refactor memory modules so that all netty code is in netty-memory.
- All code being injected into Netty packages and have dependencies on Netty's package-private code go into a new module memory-netty-buffer-patch.
- The surefire command line has been changed to allow reflection on unsafe to be used by arrow-memory-core
- Add module-info files for arrow-memory-core and arrow-memory-unsafe

### Are these changes tested?
Yes, existing tests work when run with modules.

### Are there any user-facing changes?
Yes, users now need to put --add-opens=java.base/java.nio=org.apache.arrow.memory.core,ALL-UNNAMED on their java command line instead of --add-opens=java.base/java.nio=ALL-UNNAMED since memory-core is now a named module instead of unnamed.

 **This PR includes breaking changes to public APIs.** 
Yes, users now need to put --add-opens=java.base/java.nio=org.apache.arrow.memory.core,ALL-UNNAMED on their java command line instead of --add-opens=java.base/java.nio=ALL-UNNAMED since memory-core is now a named module instead of unnamed.
* Closes: #38998

jduo commented:
It seems these changes are not usable without modularizing memory-core. arrow-vector makes use of Preconditions and BufferAllocator heavily. Since these are declared in an unnamed module (memory-core), the user has to add --open-reads=org.apache.arrow.vector=ALL-UNNAMED. This doesn't fail when running arrow-vector tests because surefire adds this automatically for this module but fails any other test that depends on arrow-vector.

The Preconditions class could be moved to a new utils module, but that doesn't deal with BufferAllocator which is very much part of memory-core out of necessity.

vibhatha commented on java/maven/module-info-compiler-maven-plugin/pom.xml:
```diff
@@ -0,0 +1,130 @@
+  </build>
+
+
+</project>
```
new line?

jduo replied:
Fixed now.

vibhatha commented on java/maven/module-info-compiler-maven-plugin/src/main/java/org/apache/arrow/maven/plugins/BaseModuleInfoCompilerPlugin.java:
```diff
@@ -0,0 +1,92 @@
+  }
+
+  /**
+   * Finds the first module-info.java file in the set of source directories.
```
@jduo Just for the curiosity, why are we just checking the first `module-info.java`?

jduo replied:
A JAR can only have one module-info file. We are arbitrarily taking the first. I haven't checked if the maven-compiler-plugin has the same behavior or if it errors out when there are multiple.

vibhatha replied:
Noted, thanks.

vibhatha commented on java/memory/memory-core/src/main/java/org/apache/arrow/memory/util/MemoryUtil.java:
```diff
@@ -142,7 +142,7 @@ public Object run() {
       final RuntimeException failure = new RuntimeException(
           "Failed to initialize MemoryUtil. You must start Java with " +
-              "`--add-opens=java.base/java.nio=ALL-UNNAMED` " +
+              "`--add-opens=java.base/java.nio=org.apache.arrow.memory.core,ALL-UNNAMED` " +
```
would this require a documentation update?

vibhatha replied:
should we include it to this PR or file a separate issue for that?

jduo replied:
I think it's best to do a separate PR for doc updates around this since there are a few of these across multiple PRs.

vibhatha commented on java/memory/memory-core/src/test/java/org/apache/arrow/memory/TestBaseAllocator.java:
```diff
@@ -448,73 +443,73 @@ public ArrowBuf empty() {
-    assertEquals(0, l1.getNumReleaseCalls());
-    assertEquals(0, l1.getNumChildren());
-    assertEquals(0, l1.getTotalMem());
+    Assert.assertEquals(0, l1.getNumPreCalls());
```
nit: should we make `l1` -> `listner1`. I know it is not related to this PR `11` vs `l1`, just for the readability. And may be the following names too?

jduo replied:
Done.

vibhatha commented on java/memory/memory-core/src/test/java/org/apache/arrow/memory/TestForeignAllocation.java:
```diff
@@ -23,6 +23,13 @@
 import java.util.ArrayList;
 import java.util.List;
 
+import org.apache.arrow.memory.AllocationListener;
+import org.apache.arrow.memory.AllocationOutcome;
+import org.apache.arrow.memory.ArrowBuf;
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.memory.ForeignAllocation;
+import org.apache.arrow.memory.OutOfMemoryException;
+import org.apache.arrow.memory.RootAllocator;
```
seems like these are not used.

vibhatha reviewed (commented):
@jduo I have added a few comments.

lidavidm commented on java/pom.xml:
```diff
@@ -870,7 +897,7 @@
             <artifactId>maven-surefire-plugin</artifactId>
             <configuration>
-              <argLine>--add-opens=java.base/java.nio=ALL-UNNAMED</argLine>
+              <argLine>--add-opens=java.base/java.nio=org.apache.arrow.memory.core,ALL-UNNAMED</argLine>
```
Just to be sure, do we still need ALL-UNNAMED?

jduo replied:
Yes, it's still needed. Netty utilizes it for example.

lidavidm commented on java/memory/memory-unsafe/src/test/java/org/apache/arrow/memory/unsafe/TestUnsafeAllocationManager.java:
```diff
@@ -15,20 +15,25 @@
 
-  private BaseAllocator createUnsafeAllocator() {
-    return new RootAllocator(BaseAllocator.configBuilder().allocationManagerFactory(UnsafeAllocationManager.FACTORY)
+  private RootAllocator createUnsafeAllocator() {
```
Ideally we never refer to RootAllocator (or BaseAllocator) as a type, only BufferAllocator

jduo replied:
Changed to return BufferAllocator instead.

lidavidm commented on java/memory/memory-core/pom.xml:
```diff
@@ -54,6 +53,30 @@
 
   <profiles>
+    <profile>
+      <id>error-prone-jdk11+</id>
```
how is this profile related to ErrorProne?

jduo replied:
This profile is really the "default" build profile for JDK9+. It doesn't relate to error-prone specifically. Would it be better to have a separate profile with the same activation?

lidavidm replied:
Ah, then just rename it so it's clear - the "error-prone" profiles we have are meant for running the ErrorProne static analyzer

jduo replied:
Changed the profile to be jdk11+.

lidavidm commented:
If we're going to include this with Arrow 15, we should get on the doc changes immediately.

jduo commented:
> If we're going to include this with Arrow 15, we should get on the doc changes immediately.

OK, I have added the doc change to this PR instead.

lidavidm commented:
@github-actions crossbow submit *java* *verification*

lidavidm commented:
Sorry, I didn't get around to this and the code freeze is today...this might have to slip to arrow 16 since I realized we never checked the other Java builds

lidavidm commented:
@github-actions crossbow submit *java* *verify*

lidavidm commented:
It appears Javadocs builds fail:

```
Error:  Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:2.10.3:jar (attach-javadocs) on project arrow-memory-core: MavenReportException: Error while generating Javadoc: 
Error:  Exit code: 1 - /Users/runner/work/crossbow/crossbow/arrow/java/memory/memory-core/src/main/java/module-info.java:18: error: class, interface, or enum expected
Error:  module org.apache.arrow.memory.core {
Error:  ^
```

jduo commented:
> Sorry, I didn't get around to this and the code freeze is today...this might have to slip to arrow 16 since I realized we never checked the other Java builds

No problem. It is probably best this gets completed in the same release as #39001 too so this gives a bit more time to examine that as well.

jduo commented:
@github-actions crossbow submit *java*

jduo commented:
@github-actions crossbow submit *java* *verify*

jduo commented:
@github-actions crossbow submit *java* *verify*

jduo commented:
@github-actions crossbow submit *java* *verify*

jduo commented:
@github-actions crossbow submit *java* *verify*

jduo commented:
@lidavidm , the javadoc generation failure is now fixed. I updated the maven-javadoc-plugin to a newer version that properly let you exclude files and excluded the module-info.java files.

jduo commented:
> It appears Javadocs builds fail:
> 
> ```
> Error:  Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:2.10.3:jar (attach-javadocs) on project arrow-memory-core: MavenReportException: Error while generating Javadoc: 
> Error:  Exit code: 1 - /Users/runner/work/crossbow/crossbow/arrow/java/memory/memory-core/src/main/java/module-info.java:18: error: class, interface, or enum expected
> Error:  module org.apache.arrow.memory.core {
> Error:  ^
> ```

This issue is now fixed by updating maven-javadoc-plugin to a version that allows source exclusions properly and excluding module-info.java files.

lidavidm commented on java/pom.xml:
```diff
@@ -776,6 +787,11 @@
             </reports>
           </reportSet>
         </reportSets>
+        <configuration>
+          <sourceFileExcludes>
+            <sourceFileExclude>**/module-info.java</sourceFileExclude>
+          </sourceFileExcludes>
+        </configuration>
```
We may want to document module-info.java in Java 9+ right?

jduo replied:
Yes, I think we should wait until JDK8 support is dropped to do this though. We run into similar issues as we ran into while compiling (need to set release to 9, but that breaks how we utilize Unsafe).

lidavidm commented on java/memory/memory-core/src/test/java/org/apache/arrow/memory/CountingAllocationListener.java:
```diff
@@ -58,6 +64,10 @@ public void onAllocation(long size) {
   @Override
   public boolean onFailedAllocation(long size, AllocationOutcome outcome) {
     if (expandOnFail) {
+      if (expandAlloc == null) {
```
Could possibly use Preconditions.checkState for this (since it's been annotated such that CheckerFramework should understand it)
